// queueable/dist/module/adapters/Channel.js
import Deferred from '../Deferred';
import LinkedQueue from '../LinkedQueue';
import { doneResult } from '../common';
import fromDom from '../fromDom';
import fromEmitter from '../fromEmitter';
/**
 * Balances a push queue with a pull queue.
 */
export default class Channel {
    constructor(pushLimit = 0, pullLimit = 0) {
        /** Determines whether new values can be pushed or pulled */
        this.closed = false;
        this.pushBuffer = new LinkedQueue(pushLimit);
        this.pullBuffer = new LinkedQueue(pullLimit);
    }
    /**
     * Pull a promise of the next [[Result]].
     */
    next() {
        if (this.closed) {
            return Promise.resolve(doneResult);
        }
        if (this.pushBuffer.length === 0) {
            const defer = new Deferred();
            // Buffer the pull to be resolved later
            this.pullBuffer.enqueue(defer);
            // Return the buffered promise that will be resolved and dequeued when a value is pushed
            return defer.promise;
        }
        const { result, defer } = this.pushBuffer.dequeue();
        defer.resolve(result);
        if (result.done) {
            this.close();
        }
        return defer.promise;
    }
    /**
     * Push the next [[Result]] value.
     *
     * @param value
     * @param done If true, closes the balancer when this result is resolved
     * @throws Throws if the balancer is already closed
     */
    push(value, done = false) {
        if (this.closed) {
            return Promise.resolve(doneResult);
        }
        const result = {
            value,
            done,
        };
        if (this.pullBuffer.length > 0) {
            return this.pullBuffer.dequeue().resolve(result);
        }
        const defer = new Deferred();
        this.pushBuffer.enqueue({ result, defer });
        return defer.promise;
    }
    /**
     * Returns itself, since [[Balancer]] already implements the iterator protocol.
     */
    [Symbol.asyncIterator]() {
        return this;
    }
    /**
     * Closes the balancer; clears the queues and makes [[Balancer.next]] only
     * return [[closedResult]].
     *
     * @param value The result value to be returned
     */
    async return(value) {
        this.close();
        return {
            done: true,
            value: value,
        };
    }
    close() {
        if (this.closed) {
            return;
        }
        this.closed = true;
        // Clear the queues
        this.pushBuffer.forEach(({ defer: { resolve } }) => void resolve(doneResult));
        this.pushBuffer.clear();
        this.pullBuffer.forEach(({ resolve }) => void resolve(doneResult));
        this.pullBuffer.clear();
    }
    /**
     * Convert [[Balancer]] to a generic async iterable iterator to hide implementation details.
     *
     * @param onReturn Optional callback for when the iterator is closed with [[Balancer.return]]
     * @throws Throws if called when closed
     */
    wrap(onReturn) {
        if (this.closed) {
            throw Error('Balancer is closed');
        }
        return {
            [Symbol.asyncIterator]() {
                return this;
            },
            next: () => this.next(),
            return: async (value) => {
                if (onReturn) {
                    onReturn();
                }
                return this.return(value);
            },
        };
    }
}
Channel.fromDom = fromDom(() => new Channel());
Channel.fromEmitter = fromEmitter(() => new Channel());
