// queueable/dist/lib/adapters/LastResult.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Deferred_1 = __importDefault(require("../Deferred"));
const common_1 = require("../common");
const fromDom_1 = __importDefault(require("../fromDom"));
const fromEmitter_1 = __importDefault(require("../fromEmitter"));
class LastResult {
    constructor() {
        this.buffer = new Deferred_1.default();
        this.closed = false;
        this.resolved = false;
        this.requested = false;
    }
    push(value, done = false) {
        if (this.closed) {
            throw Error('Iterator closed');
        }
        const result = {
            value,
            done,
        };
        if (this.resolved === false) {
            this.resolved = true;
        }
        else {
            this.buffer = new Deferred_1.default();
            this.resolved = false;
        }
        this.requested = false;
        this.buffer.resolve(result);
        return this.buffer.promise;
    }
    async next() {
        if (this.closed) {
            return common_1.doneResult;
        }
        this.requested = true;
        return this.buffer.promise;
    }
    async return(value) {
        this.closed = true;
        if (!this.resolved && this.requested) {
            this.buffer.resolve(common_1.doneResult);
        }
        return Promise.resolve({
            value: value,
            done: true,
        });
    }
    wrap(onReturn) {
        const wrapped = {
            next: () => this.next(),
            [Symbol.asyncIterator]() {
                return this;
            },
            return: (value) => {
                if (onReturn) {
                    onReturn();
                }
                return this.return(value);
            },
        };
        return wrapped;
    }
    [Symbol.asyncIterator]() {
        return this;
    }
}
exports.default = LastResult;
LastResult.fromDom = fromDom_1.default(() => new LastResult());
LastResult.fromEmitter = fromEmitter_1.default(() => new LastResult());
