// @textile/multiaddr/dist/codec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.protoFromTuple = exports.ParseError = exports.cleanPath = exports.isValidBuffer = exports.fromBuffer = exports.fromString = exports.stringToBuffer = exports.bufferToString = exports.bufferToTuples = exports.sizeForAddr = exports.tuplesToBuffer = exports.tuplesToStringTuples = exports.stringTuplesToTuples = exports.stringTuplesToString = exports.stringToStringTuples = void 0;
/* eslint-disable @typescript-eslint/no-use-before-define */
const varint_1 = __importDefault(require("varint"));
const convert = __importStar(require("./convert"));
const protocols_1 = require("./protocols");
// string -> [[str name, str addr]... ]
function stringToStringTuples(str) {
    const tuples = [];
    const parts = str.split('/').slice(1); // skip first empty elem
    if (parts.length === 1 && parts[0] === '') {
        return [];
    }
    for (let p = 0; p < parts.length; p++) {
        const part = parts[p];
        const proto = protocols_1.protocols(part);
        if (proto.size === 0) {
            tuples.push([part]);
            continue;
        }
        p++; // advance addr part
        if (p >= parts.length) {
            throw ParseError('invalid address: ' + str);
        }
        // if it's a path proto, take the rest
        if (proto.path) {
            tuples.push([
                part,
                // TODO: should we need to check each path part to see if it's a proto?
                // This would allow for other protocols to be added after a unix path,
                // however it would have issues if the path had a protocol name in the path
                cleanPath(parts.slice(p).join('/')),
            ]);
            break;
        }
        tuples.push([part, parts[p]]);
    }
    return tuples;
}
exports.stringToStringTuples = stringToStringTuples;
// [[str name, str addr]... ] -> string
function stringTuplesToString(tuples) {
    const parts = [];
    tuples.map(tup => {
        const proto = protoFromTuple(tup);
        parts.push(proto.name);
        if (tup.length > 1) {
            parts.push(tup[1]);
        }
    });
    return cleanPath(parts.join('/'));
}
exports.stringTuplesToString = stringTuplesToString;
// [[str name, str addr]... ] -> [[int code, Buffer]... ]
function stringTuplesToTuples(tuples) {
    return tuples.map(tup => {
        if (!Array.isArray(tup)) {
            tup = [tup];
        }
        const proto = protoFromTuple(tup);
        if (tup.length > 1) {
            return [proto.code, convert.toBuffer(proto.code, tup[1])];
        }
        return [proto.code];
    });
}
exports.stringTuplesToTuples = stringTuplesToTuples;
// [[int code, Buffer]... ] -> [[str name, str addr]... ]
function tuplesToStringTuples(tuples) {
    return tuples.map(tup => {
        const proto = protoFromTuple(tup);
        if (tup.length > 1) {
            return [proto.code, convert.toString(proto.code, tup[1])];
        }
        return [proto.code];
    });
}
exports.tuplesToStringTuples = tuplesToStringTuples;
// [[int code, Buffer ]... ] -> Buffer
function tuplesToBuffer(tuples) {
    return fromBuffer(Buffer.concat(tuples.map(tup => {
        const proto = protoFromTuple(tup);
        let buf = Buffer.from(varint_1.default.encode(proto.code));
        if (tup.length > 1) {
            buf = Buffer.concat([buf, tup[1]]); // add address buffer
        }
        return buf;
    })));
}
exports.tuplesToBuffer = tuplesToBuffer;
function sizeForAddr(p, addr) {
    if (p.size > 0) {
        return p.size / 8;
    }
    else if (p.size === 0) {
        return 0;
    }
    else {
        const size = varint_1.default.decode(addr);
        return size + varint_1.default.decode.bytes;
    }
}
exports.sizeForAddr = sizeForAddr;
// Buffer -> [[int code, Buffer ]... ]
function bufferToTuples(buf) {
    const tuples = [];
    let i = 0;
    while (i < buf.length) {
        const code = varint_1.default.decode(buf, i);
        const n = varint_1.default.decode.bytes;
        const p = protocols_1.protocols(code);
        const size = sizeForAddr(p, buf.slice(i + n));
        if (size === 0) {
            tuples.push([code]);
            i += n;
            continue;
        }
        const addr = buf.slice(i + n, i + n + size);
        i += size + n;
        if (i > buf.length) {
            // did not end _exactly_ at buffer.length
            throw ParseError('Invalid address buffer: ' + buf.toString('hex'));
        }
        // ok, tuple seems good.
        tuples.push([code, addr]);
    }
    return tuples;
}
exports.bufferToTuples = bufferToTuples;
// Buffer -> String
function bufferToString(buf) {
    const a = bufferToTuples(buf);
    const b = tuplesToStringTuples(a);
    return stringTuplesToString(b);
}
exports.bufferToString = bufferToString;
// String -> Buffer
function stringToBuffer(str) {
    str = cleanPath(str);
    const a = stringToStringTuples(str);
    const b = stringTuplesToTuples(a);
    return tuplesToBuffer(b);
}
exports.stringToBuffer = stringToBuffer;
// String -> Buffer
function fromString(str) {
    return stringToBuffer(str);
}
exports.fromString = fromString;
// Buffer -> Buffer
function fromBuffer(buf) {
    const err = validateBuffer(buf);
    if (err)
        throw err;
    return Buffer.from(buf); // copy
}
exports.fromBuffer = fromBuffer;
function validateBuffer(buf) {
    try {
        bufferToTuples(buf); // try to parse. will throw if breaks
    }
    catch (err) {
        return err;
    }
}
function isValidBuffer(buf) {
    return validateBuffer(buf) === undefined;
}
exports.isValidBuffer = isValidBuffer;
function cleanPath(str) {
    return ('/' +
        str
            .trim()
            .split('/')
            .filter(a => a)
            .join('/'));
}
exports.cleanPath = cleanPath;
function ParseError(str) {
    return new Error('Error parsing address: ' + str);
}
exports.ParseError = ParseError;
function protoFromTuple(tup) {
    const proto = protocols_1.protocols(tup[0]);
    return proto;
}
exports.protoFromTuple = protoFromTuple;
